### PR TITLE
Sqlite3 Port

### PR DESCRIPTION
**Pull request by Ruide**
This is a port of sqlite3. Now we can use FFI to take advantage of sqlite3 in Rust 👯‍♀️ .

The README.md file explains the context also.

All changes are under samplecode/sqlite3.

### CODE REVIEW

**Comment by mssun** [comment]
Hi, I found that you introduce some ocalls to help your porting:

```
        void ocall_print_error([in, string] const char *str);
        void ocall_print_string([in, string] const char *str);
        void ocall_println_string([in, string] const char *str);

        int ocall_lstat([in, string] const char *path, [in, out, size=size] struct stat *buf, size_t size) propagate_errno;
        int ocall_stat([in, string] const char *path, [in, out, size=size] struct stat *buf, size_t size);
        int ocall_fstat(int fd, [in, out, size=size] struct stat* buf, size_t size);
        int ocall_ftruncate(int fd, off_t length);

        char* ocall_getcwd([out, size=size] char *buf, size_t size) propagate_errno;
        int ocall_getpid(void);
        int ocall_getuid(void);
        char* ocall_getenv([in, string] const char *name);

        int ocall_open64([in, string] const char *filename, int flags, mode_t mode);
        int ocall_close(int fd);
        off_t ocall_lseek64(int fd, off_t offset, int whence) propagate_errno;
        int ocall_read(int fd, [out, size=count] void *buf, size_t count) propagate_errno;
        int ocall_write(int fd, [in, size=count] const void *buf, size_t count) propagate_errno;
        int ocall_fsync(int fd);
        int ocall_fcntl(int fd, int cmd, [in, size=size] void* arg, size_t size) propagate_errno;
        int ocall_unlink([in, string] const char *pathname);
        int ocall_fchmod(int fd, mode_t mode);
        int ocall_fchown(int fd, uid_t owner, gid_t group);
```

Seems like you miss something of the threat model we face in SGX. When you use `ocall_write` to write data into the disk, it's still plaintext. There is no integrity and confidentiality for the database implementation.

If we want to provide a sample code of database in SGX, we need to rethink the interfaces between trusted and untrusted world. For file IO, we should consider using protected FS. For other unrelated APIs, we can remove them. We already have a level-db implementation in the Teaclave platform: https://github.com/apache/incubator-teaclave/tree/master/common.

File: samplecode/sqlite3/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# Welcome to RUST-SGX-SQLITE SampleCode!
+
+Hi! This is a port of **Sqlite3** to **Rust-SGX-SDK**. This port reuses a lot ofcode and contents in https://github.com/yerzhan7/SGX_SQLite. Special thanks to @yerzhan7 for open sourcing his/her work! Since this port borrows a lot of content from SGX_SQLite, so they shares a lot of common properties (This port literally does all the things SGX_SQLite does!). I cite @yerzhan7's intro in the following.
+
+> SQLite database inside a secure Intel SGX enclave (Linux).
+
+> You can execute your SQL statements securely. The official SQLite library ("sqlite.c", "sqlite.h") is entirely loaded into an enclave. However, if you want to save your database (i.e. not in-memory database) it will save *.db file without any encryption yet. (need to implement data sealing in the future)
```
Comment (mssun) [inline samplecode/sqlite3/README.md:7]: Ok, I saw your comment here. This does explain my previous question.

**Comment by Ruide** [comment]
> Hi, I found that you introduce some ocalls to help your porting:
> 
> ```
>         void ocall_print_error([in, string] const char *str);
>         void ocall_print_string([in, string] const char *str);
>         void ocall_println_string([in, string] const char *str);
> 
>         int ocall_lstat([in, string] const char *path, [in, out, size=size] struct stat *buf, size_t size) propagate_errno;
>         int ocall_stat([in, string] const char *path, [in, out, size=size] struct stat *buf, size_t size);
>         int ocall_fstat(int fd, [in, out, size=size] struct stat* buf, size_t size);
>         int ocall_ftruncate(int fd, off_t length);
> 
>         char* ocall_getcwd([out, size=size] char *buf, size_t size) propagate_errno;
>         int ocall_getpid(void);
>         int ocall_getuid(void);
>         char* ocall_getenv([in, string] const char *name);
> 
>         int ocall_open64([in, string] const char *filename, int flags, mode_t mode);
>         int ocall_close(int fd);
>         off_t ocall_lseek64(int fd, off_t offset, int whence) propagate_errno;
>         int ocall_read(int fd, [out, size=count] void *buf, size_t count) propagate_errno;
>         int ocall_write(int fd, [in, size=count] const void *buf, size_t count) propagate_errno;
>         int ocall_fsync(int fd);
>         int ocall_fcntl(int fd, int cmd, [in, size=size] void* arg, size_t size) propagate_errno;
>         int ocall_unlink([in, string] const char *pathname);
>         int ocall_fchmod(int fd, mode_t mode);
>         int ocall_fchown(int fd, uid_t owner, gid_t group);
> ```
> 
> Seems like you miss something of the threat model we face in SGX. When you use `ocall_write` to write data into the disk, it's still plaintext. There is no integrity and confidentiality for the database implementation.
> 
> If we want to provide a sample code of database in SGX, we need to rethink the interfaces between trusted and untrusted world. For file IO, we should consider using protected FS. For other unrelated APIs, we can remove them. We already have a level-db implementation in the Teaclave platform: https://github.com/apache/incubator-teaclave/tree/master/common.

Thanks for your quick response! Yes, indeed you are right, I used several ocall for resolving and proxy syscall inside enclave. And yes, thank you for noticing my comments for the untrusted ocall of file io in the README file! Do you have any suggestions on how you would like me to proceed? I think I could replace file io with protected-fs. Or do you have any other thoughts? Thanks for your comments again!

File: samplecode/sqlite3/enclave/Enclave.edl [hunk 2]
```diff
@@ -32,27 +32,6 @@ enclave {
...
-        int ocall_open64([in, string] const char *filename, int flags, mode_t mode);
-        int ocall_close(int fd);
-        off_t ocall_lseek64(int fd, off_t offset, int whence) propagate_errno;
-        int ocall_read(int fd, [out, size=count] void *buf, size_t count) propagate_errno;
-        int ocall_write(int fd, [in, size=count] const void *buf, size_t count) propagate_errno;
-        int ocall_fsync(int fd);
-        int ocall_fcntl(int fd, int cmd, [in, size=size] void* arg, size_t size) propagate_errno;
-        int ocall_unlink([in, string] const char *pathname);
-        int ocall_fchmod(int fd, mode_t mode);
-        int ocall_fchown(int fd, uid_t owner, gid_t group);
```
Comment (Ruide) [inline samplecode/sqlite3/enclave/Enclave.edl]: Clean up all the ocall syscalls for security concern mentioned by @mssun. Although only support in-memory SQLite database now.

**Comment by mssun** [comment]
I found that you introduce some headers, what's the purpose? They are in LGPL license which are not accepted [1] by the Apache projects.

[1] https://www.apache.org/legal/resolved.html

I don't think this can be included in Teaclave SGX SDK as samples.

- The purpose of samples in Teaclave SGX SDK is to demonstrate the usages of SDK or the benefits of using Rust. However, most of the code are written in C. I'm not convinced to include this in the samples.
- The license issue should be addressed. As I said, I found some LGPL headers which are not accepted. Also, you mentioned that this sqlite implementation is from another project, we also need to make sure the license are accepted.
- This PR introduce a lot of code which are hard to audit/review. Also, most of code are not related with the SDK itself. Although this is sample code, we still need to ensure the security in case other people grab it without auditing it. IIRC, there was a peer-reviewed paper include a bug in our sample code as their findings.

Overall, I hope you can understand my concerns.

BTW, it's totally okay to provide it as a separate project instead of merging into this repo. You can create a new repo by yourself and open source it. Or, we have an organization called Secure Computing Community (https://github.com/sccommunity), which hosts related (experimental) projects and third-party libraries around Teaclave. I think that's a good place to put your contribution.

Feel free to comment if you have different ideas. Thanks.

**Comment by Ruide** [comment]
> I found that you introduce some headers, what's the purpose? They are in LGPL license which are not accepted [1] by the Apache projects.
> 
> [1] https://www.apache.org/legal/resolved.html
> 

They are system headers. The common/inc folder misses some of the definitions. They could be trimmed to meet the licensing requirement? (I am not an expert with Licensing)

> I don't think this can be included in Teaclave SGX SDK as samples.
> 
> * The purpose of samples in Teaclave SGX SDK is to demonstrate the usages of SDK or the benefits of using Rust. However, most of the code are written in C. I'm not convinced to include this in the samples.

I use the upstream sqlite c source code with an addition of about 30 lines of code to resolve symbols. To avoid the amount of c code from preventing review, it could be simplified to a fetch and patch.

> * The license issue should be addressed. As I said, I found some LGPL headers which are not accepted. Also, you mentioned that this sqlite implementation is from another project, we also need to make sure the license are accepted.

My phrases were a little too loose. As I mentioned above, mostly is upstream with 30 lines of additions.

> * This PR introduce a lot of code which are hard to audit/review. Also, most of code are not related with the SDK itself. Although this is sample code, we still need to ensure the security in case other people grab it without auditing it. IIRC, there was a peer-reviewed paper include a bug in our sample code as their findings.
> 

Yea, I remember that paper, `TeeRex` right. Yea, totally agreed. It would be easy to audit if it is changed to a fetch and patch. Although we need to trust upstream sqlite for this case (I find this a reasonable assumption?).

> Overall, I hope you can understand my concerns.
> 

Sure, totally understood. I am closing this PR. 

> BTW, it's totally okay to provide it as a separate project instead of merging into this repo. You can create a new repo by yourself and open source it. Or, we have an organization called Secure Computing Community (https://github.com/sccommunity), which hosts related (experimental) projects and third-party libraries around Teaclave. I think that's a good place to put your contribution.
> 
> Feel free to comment if you have different ideas. Thanks.

I think in general, this is a good learning experience of Rust SGX SDK for me. And now I have a deeper understanding of how most of intel sgx sdk is not fully rewritten in Rust but linked to by FFI. And how the parts are piled together.

But I do think by limiting the enclave programming language to Rust, and not trusting any external battlefield-tested library (sqlite in this case) would limit the scope of this project. I assume for a long time, there would not be a full Rust rewrite for sqlite. I think letting people know Rust SGX SDK not only allows embedded Rust crates but also can incorporate external code is important (Since in the current state, Rust has not yet rewritten everything, hopefully in the future, it would :D). And I do think arguing for better security by limiting the addition of new functionality shall be reconsidered. Since a better way could be to enforce strong auditing on the added code (tls-client or sqlite in this case). For example, we could audit the changes to upstream sqlite (Sorry for not making it easy to review!). And we could argue for the security boundary, e.g., not responsible for bugs in external libraries.

Anyway, thank you so much for your time and great comments! They make a lot of sense and much appreciated!

**Comment by mssun** [comment]
> But I do think by limiting the enclave programming language to Rust, and not trusting any external battlefield-tested library (sqlite in this case) would limit the scope of this project.

Yes, I agree with you. This is another topic we can discuss more.

Originally, the SGX SDK is designed to serve logics (app and enclave) written in Rust code. For example, ocall interfaces are only for Rust standard library. However, if you want to use FFI to call C library like sqlite, you have to implement some new ocall interfaces, which is extremely dangerous. However, we can provide an "ocall" abstraction layer (or so called platform abstraction layer) and give Rust and C interfaces (ABIs) to serve both Rust standard library and libc. This layer only focus on providing a secure yet universal "ocall" interfaces. Specifically, this layer is written in Rust and export C FFI to serve upper C libraries. I understand my description is still very obscure, but that's the rough idea I have. Here is a simple graph to demonstrate my thoughts:

```
--------------------------------------------------------------------------
Enclave (Rust)        | Enclave (C)
----------------------+-------------+-------------------------------------
Rust standard library | libc        | SGX platform interfaces (Rust/C FFI)
----------------------+-------------+
ocall interfaces (Rust)             |
------------------------------------+-------------------------------------
```

**Comment by Ruide** [comment]
> > But I do think by limiting the enclave programming language to Rust, and not trusting any external battlefield-tested library (sqlite in this case) would limit the scope of this project.
> 
> Yes, I agree with you. This is another topic we can discuss more.
> 
> Originally, the SGX SDK is designed to serve logics (app and enclave) written in Rust code. For example, ocall interfaces are only for Rust standard library. However, if you want to use FFI to call C library like sqlite, you have to implement some new ocall interfaces, which is extremely dangerous. 

I believe I have got rid of all the newly created ocall interfaces for sqlite in my previous commit? I believe it is now using the syscalls provided by tlibc (from Intel SGX SDK I guess). I left a comment before. Please Correct me if I am wrong!

> However, we can provide an "ocall" abstraction layer (or so called platform abstraction layer) and give Rust and C interfaces (ABIs) to serve both Rust standard library and libc. This layer only focus on providing a secure yet universal "ocall" interfaces. Specifically, this layer is written in Rust and export C FFI to serve upper C libraries. I understand my description is still very obscure, but that's the rough idea I have. Here is a simple graph to demonstrate my thoughts:
> 

Yea, I remember there was also a paper from Singapore doing such a thing. If I'm understanding correctly, you want to use Rust to extend the current Rust sgx std library and support all symbols provided in Intel's tlibc (by replacing the implementations with Rust and export the symbols with C ffi)? So that we won't need tlibc anymore. And of cause, there would be security checks written in Rust.  

> ```
> --------------------------------------------------------------------------
> Enclave (Rust)        | Enclave (C)
> ----------------------+-------------+-------------------------------------
> Rust standard library | libc        | SGX platform interfaces (Rust/C FFI)
> ----------------------+-------------+
> ocall interfaces (Rust)             |
> ------------------------------------+-------------------------------------
> ```

Thanks for drawing the figure! So my understanding of this figure is that, Enclave written in Rust fully runs on top of Rust sgx_tstd. But Enclave written in C is not taking advantage of rigorous checking of Rust sgx_tstd but instead relies on Intel's tlibc (and Intel does not provide checks on the primitive ocalls? I believe they also do?) So if I am understanding correctly, you are saying that Rust sgx_tstd is more trustworthy than tlibc provided by Intel? So we shall not use external C libraries even if they do not bring further ocalls? Please correct me if I am wrong! Thank you so much for your explanations!

**Comment by mssun** [comment]
Sorry, it seems that my reply brings you more questions.

> But Enclave written in C is not taking advantage of rigorous checking of Rust sgx_tstd but instead relies on Intel's tlibc (and Intel does not provide checks on the primitive ocalls? I believe they also do?)

I think there's one misunderstanding: Intel's tlibc does not provide any APIs which needs ocalls. In comparison, `sgx_tstd` provides APIs which requires ocalls (e.g., `std::net`). If someone want to use external C libraries which needs ocalls, he cannot use `sgx_std` obviously.  Then, the only thing he can do is to write their own ocalls to extend Intel's tlibc, which may be vulnerable [1].

The solution I mentioned is to provide these extended libc API on top of the ocall interfaces written in Rust which have better checking and auditing. Hope this can explain my previous post.

```
----------------------------------------------------------------------------
Enclave (Rust)        | Enclave (C)
----------------------+---------------+-------------------------------------
Rust standard library | extended libc | SGX platform interfaces (Rust/C FFI)
----------------------+---------------+
ocall interfaces (Rust)               |
--------------------------------------+-------------------------------------
```

- [1] https://github.com/openenclave/openenclave/security/advisories/GHSA-525h-wxcc-f66m

**Comment by Ruide** [comment]
Thank you for your reply, they make sense a lot!

> > But Enclave written in C is not taking advantage of rigorous checking of Rust sgx_tstd but instead relies on Intel's tlibc (and Intel does not provide checks on the primitive ocalls? I believe they also do?)
> 
> I think there's one misunderstanding: Intel's tlibc does not provide any APIs which needs ocalls. In comparison, `sgx_tstd` provides APIs which requires ocalls (e.g., `std::net`). If someone want to use external C libraries which needs ocalls, he cannot use `sgx_std` obviously. Then, the only thing he can do is to write their own ocalls to extend Intel's tlibc, which may be vulnerable [1].
> 

Thanks for your correction! Yea, I originally thought tlibc provide some ocall to support some system calls. In this case, if enclave code only depends on tlibc, it won't leave the enclave. Then I guess the [current version of SQLite port](https://github.com/apache/incubator-teaclave-sgx-sdk/pull/274/commits/9cdf6c3f86b4631b51eef4802adc79e13cf36f87) won't leave enclave during execution (since I make it support in-memory database only and it currently does not introduce any additional ocalls)? Please correct me if I am wrong! 

> The solution I mentioned is to provide these extended libc API on top of the ocall interfaces written in Rust which have better checking and auditing. Hope this can explain my previous post.
> 
> ```
> ----------------------------------------------------------------------------
> Enclave (Rust)        | Enclave (C)
> ----------------------+---------------+-------------------------------------
> Rust standard library | extended libc | SGX platform interfaces (Rust/C FFI)
> ----------------------+---------------+
> ocall interfaces (Rust)               |
> --------------------------------------+-------------------------------------
> ```
> 
> * [1] [GHSA-525h-wxcc-f66m](https://github.com/openenclave/openenclave/security/advisories/GHSA-525h-wxcc-f66m)

Yes, this feels like a standard solution similar to SGX SDK solutions like Google Asylo, MS Openenclave ([link here](https://github.com/openenclave/openenclave/blob/116edfdb863c4b51cb00807ee5d416f2dfa9e3c9/include/openenclave/edl/syscall.edl)) or libOS solutions like MAYI Occlum and others. Thanks for your explanation! And I guess although this might make people recklessly port libraries, this still might be a good option to have.

**Comment by mssun** [comment]
> Then I guess the current version of SQLite port won't leave enclave during execution (since I make it support in-memory database only and it currently does not introduce any additional ocalls)?

Yes. Your workaround is to give some non-functional stubs. This will still cause runtime errors. The best way is to trim those functions and APIs, and providing a customized SQLite only for SGX environment.

**Comment by Ruide** [comment]
> > Then I guess the current version of SQLite port won't leave enclave during execution (since I make it support in-memory database only and it currently does not introduce any additional ocalls)?
> 
> Yes. Your workaround is to give some non-functional stubs. This will still cause runtime errors. The best way is to trim those functions and APIs, and providing a customized SQLite only for SGX environment.

Yes, I did and yes it will give "Error: no ocall implementation for xxx" at runtime when called. You are right, I guess I would need to replace these stubs with idiomatic enclave primitives. Thanks for your clarification!

**Pull request closed without merging**